cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/executor/Executor.java

```diff
@@ -68,107 +73,140 @@ public Executor(KafkaCruiseControlConfig config, Time time, MetricRegistry dropw
     _statusCheckingIntervalMs = config.getLong(KafkaCruiseControlConfig.EXECUTION_PROGRESS_CHECK_INTERVAL_MS_CONFIG);
     _proposalExecutor =
         Executors.newSingleThreadExecutor(new KafkaCruiseControlThreadFactory("ProposalExecutor", false, LOG));
-    _state = new AtomicReference<>(ExecutorState.State.NO_TASK_IN_PROGRESS);
-    _stopRequested = false;
+    _state = ExecutorState.State.NO_TASK_IN_PROGRESS;
+    _stopRequested = new AtomicBoolean(false);
+    _stateLock = new ReentrantLock();
   }
 
   /**
    * Check whether the executor is executing a set of proposals.
    */
   public ExecutorState state() {
-    switch (_state.get()) {
-      case NO_TASK_IN_PROGRESS:
-        return ExecutorState.noTaskInProgress();
-      case EXECUTION_STARTED:
-        return ExecutorState.executionStarted();
-      case LEADER_MOVEMENT_TASK_IN_PROGRESS:
-        return ExecutorState.leaderMovementInProgress();
-      case REPLICA_MOVEMENT_TASK_IN_PROGRESS:
-        ExecutorState currState =
-            ExecutorState.replicaMovementInProgress(_numFinishedPartitionMovements,
-                                                    _executionTaskManager.remainingPartitionMovements(),
-                                                    _executionTaskManager.inProgressTasks(),
-                                                    _executionTaskManager.abortingTasks(),
-                                                    _executionTaskManager.abortedTasks(),
-                                                    _executionTaskManager.deadTasks(),
-                                                    _executionTaskManager.remainingDataToMoveInMB(),
-                                                    _finishedDataMovementInMB);
-        if (_state.get() == ExecutorState.State.REPLICA_MOVEMENT_TASK_IN_PROGRESS) {
-          return currState;
-        } else {
-          return state();
-        }
-      case STOPPING_EXECUTION:
-        return ExecutorState.stopping(_numFinishedPartitionMovements,
-                                      _executionTaskManager.remainingPartitionMovements(),
-                                      _executionTaskManager.inProgressTasks(),
-                                      _executionTaskManager.abortingTasks(),
-                                      _executionTaskManager.abortedTasks(),
-                                      _executionTaskManager.deadTasks(),
-                                      _executionTaskManager.remainingDataToMoveInMB(),
-                                      _finishedDataMovementInMB);
-      default:
-        throw new IllegalStateException("Should never be here!");
+    _stateLock.lock();
```

Comment:
The locking becomes a little ugly here. Given that the main thing we want to protect is essentially the state query consistency, maybe we can just create an executor state cache object. The state cache will only be updated by the ExecutionRunnable atomically. Hence we don't need to worry about the consistency.